From edf5699c189bf8da642297fe327e19a6c7674091 Mon Sep 17 00:00:00 2001
From: Kevin McCarthy <kevin@8t8.us>
Date: Sat, 28 Dec 2019 15:43:07 -0800
Subject: Fix crash when polling a closed ssl connection

Commit 8353407c enabled checking for buffered OpenSSL/GnuTLS data when
polling, but neglected to check if the connection was already closed.

This can be triggered during imap_logout() if the connection write of
"LOGOUT" fails and closes the connection, before the poll.  It's a bit
tricky to trigger because imap_logout_all() checks for a closed
connection, so the failure needs to take place at that last write.

Thanks to Stefan Sperling for pointing out the problem, complete with
a backtrace and patch.  (This commit takes a different approach for a
stable-branch fix.)

Upstream-commit: https://gitlab.com/muttmua/mutt/commit/edf5699c189bf8da642297fe327e19a6c7674091
Co-authored-by: Richard Russon <rich@flatcap.org>
---
 mutt_ssl.c        | 3 +++
 mutt_ssl_gnutls.c | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/mutt_ssl.c b/mutt_ssl.c
index 3dee74c6..6978e4e4 100644
--- a/mutt_ssl.c
+++ b/mutt_ssl.c
@@ -465,6 +465,9 @@ static int ssl_socket_poll (CONNECTION* conn, time_t wait_secs)
 {
   sslsockdata *data = conn->sockdata;
 
+  if (!data)
+    return -1;
+
   if (SSL_has_pending (data->ssl))
     return 1;
   else
diff --git a/mutt_ssl_gnutls.c b/mutt_ssl_gnutls.c
index 71873766..25c6778a 100644
--- a/mutt_ssl_gnutls.c
+++ b/mutt_ssl_gnutls.c
@@ -193,6 +193,9 @@ static int tls_socket_poll (CONNECTION* conn, time_t wait_secs)
 {
   tlssockdata *data = conn->sockdata;
 
+  if (!data)
+    return -1;
+
   if (gnutls_record_check_pending (data->state))
     return 1;
   else
-- 
2.24.1

